spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/CarryoverRemoveIterator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Iterator;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.StructType;
+
+class CarryoverRemoveIterator extends ChangelogIterator {
+  private final Iterator<Row> rowIterator;
+
+  private Row deletedRow = null;
+  private long deletedRowCount = 0;
```

Comment:
We can use a stack/list for the same purpose, but this solution has less memory cost.

Reply:
Are there any potential issues with reusing the same object given that we mutate that object in place in pre/post image computation? Or is it safe because there will be an exception if multiple DELETE rows get there? Seems like it is going to work, just checking.

Reply:
We don't mutate the object while removing carry-over rows. To provide more context, this PR split the `ChangelogIterator` to two iterators, the first one(CarryoverRemoveIterator) only handles the removing carry-over rows, after that,  we apply the second one(Modified `ChangelogIterator`), which computes updates only.

Reply:
Correct, we don't mutate them in `CarryoverRemoveIterator` but in `ChangelogIterator` when computing update images. However, that seems to happen only for a pair of delete and insert, which should work, I guess.

Reply:
That seems to work correctly.

Reply:
That's a good point. I overlooked it at the beginning. Thinking a bit more. It wouldn't be an issue in the current use case as you mentioned. The shared object only happens when there are multiple identical delete rows. For example, assume the following rows through the `removeCarryoverIterator`
```
   {0, "a", "data", DELETE}
   {0, "a", "data", DELETE}
   {3, "a", "new_data", INSERT}
```
The iterator output will be identical to the input with a subtle difference, which is the first 2 rows share the same object. It is fine in the current use case, since `computeUpdates` will throw an exception when there are duplicated delete rows. 
```
   {0, "a", "data", DELETE}
   {0, "a", "data", DELETE}
   {3, "a", "new_data", INSERT}
```

There is minor a risk. That in case if an iterator in the future concatenate the `RemoveCarryoverIterator`, and try to mutate one of a delete, this would be surprise, since it can mutate multiple rows. Overall, we are good here.